### PR TITLE
Empty RecastNavigation gem created.

### PR DESCRIPTION
**Pull request by AMZN-Olex**
- I executed `scripts/o3de.bat create-gem -gp Gems/RecastNavigation`
- I modifed `RecastNavigation\gem.json` to provide basic information for the coming gem there.

Actual code implementation will come after this PR.

RFC: https://github.com/o3de/sig-simulation/issues/10

Signed-off-by: Olex Lozitskiy <5432499+AMZN-Olex@users.noreply.github.com>

### CODE REVIEW

File: Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<RecastNavigationSystemComponent, AZ::Component>()
+                ->Version(0)
+                ;
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<RecastNavigationSystemComponent>("RecastNavigation", "[Description of functionality provided by this System Component]")
```
Comment (lmbr-pip) [inline Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp:20]: Add a description as to what this component is

File: Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+        static void Reflect(AZ::ReflectContext* context);
+
+        static void GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided);
+        static void GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible);
+        static void GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required);
+        static void GetDependentServices(AZ::ComponentDescriptor::DependencyArrayType& dependent);
+
+        RecastNavigationSystemComponent();
+        ~RecastNavigationSystemComponent();
```
Comment (lmbr-pip) [inline Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.h:26]: Nit: should this be virtual destructor? Assume that we may delete components via reference to base class.

File: Gems/RecastNavigation/gem.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    "type": "Code",
+    "summary": "Recast Navigation gem provides navigation features using Recast/Detour library that can be found at https://github.com/recastnavigation/recastnavigation",
+    "canonical_tags": [
+        "Gem"
+    ],
+    "user_tags": [
+        "Navigation"
+    ],
+    "icon_path": "preview.png",
+    "requirements": "This gem requires https://github.com/recastnavigation/recastnavigation",
```
Comment (lmbr-pip) [inline Gems/RecastNavigation/gem.json:17]: How are those requirements enforced? This reads that you should have this locally.

Comment (AMZN-Olex) [reply Gems/RecastNavigation/gem.json]: Added a clarification.

File: Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.h (hunk 2 above)
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.h]: Override added.

File: Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp (hunk 1 above)
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp]: Re-wrote the description. At the moment, this template component does nothing.

**Comment by AMZN-Olex** [comment]
> Do we want to merge the empty gem with no functionality into development? Would it not be better to have a feature branch in a fork where these kind of changes can be reviewed and then make a PR into development when the gem has the core functionality?

@AMZN-alexpete - That is a valid question. My worry is that if I push that into a temporary fork, then it would require a large PR to go into development?
The good news is that I have real code coming in right after this PR within a day or two. So the window for an empty branch would be quite small and it's in development branch, which I personally think of as a development fork.
Thoughts?

**Comment by ParasiteSt** [comment]
I agree with @AMZN-Olex with utilizing the development branch. But we do have to keep in mind a lot of people use the development branch for all the extra bug fixes, so there should be a ping on discord to let everyone know that this gem is not usable until X PR gets merged.

**Comment by AMZN-alexpete** [comment]
> > Do we want to merge the empty gem with no functionality into development? Would it not be better to have a feature branch in a fork where these kind of changes can be reviewed and then make a PR into development when the gem has the core functionality?
> 
> @AMZN-alexpete - That is a valid question. My worry is that if I push that into a temporary fork, then it would require a large PR to go into development? The good news is that I have real code coming in right after this PR within a day or two. So the window for an empty branch would be quite small and it's in development branch, which I personally think of as a development fork. Thoughts?

It certainly could be bigger - however, I found with my [Stars gem PR](https://github.com/o3de/o3de/pull/8624) the default template created a ton of boiler plate files I didn't need and I was able to delete half of them or more and get a much smaller PR than if I had just committed the default files.    You may find the same to be true - that you end up with fewer files, though the amount of lines of code will likely be larger.  When you are ready to merge the gem with basic functionality to development you could mention the PRs in the branch and just ask the original reviewers to treat it like they would a merge commit.

**Comment by AMZN-Olex** [comment]
> > > Do we want to merge the empty gem with no functionality into development? Would it not be better to have a feature branch in a fork where these kind of changes can be reviewed and then make a PR into development when the gem has the core functionality?
> > 
> > 
> > @AMZN-alexpete - That is a valid question. My worry is that if I push that into a temporary fork, then it would require a large PR to go into development? The good news is that I have real code coming in right after this PR within a day or two. So the window for an empty branch would be quite small and it's in development branch, which I personally think of as a development fork. Thoughts?
> 
> It certainly could be bigger - however, I found with my [Stars gem PR](https://github.com/o3de/o3de/pull/8624) the default template created a ton of boiler plate files I didn't need and I was able to delete half of them or more and get a much smaller PR than if I had just committed the default files. You may find the same to be true - that you end up with fewer files, though the amount of lines of code will likely be larger. When you are ready to merge the gem with basic functionality to development you could mention the PRs in the branch and just ask the original reviewers to treat it like they would a merge commit.

@AMZN-alexpete - yes, that is a very good point. However, as I have been developing ahead of this PR, I found myself needing Editor targets and unit tests targets for both non-Editor and Editor flavors. Thus, I know I will need all of those. :)

File: Gems/RecastNavigation/Code/Platform/Mac/PAL_mac.cmake [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) Contributors to the Open 3D Engine Project.
+# For complete copyright and license terms please see the LICENSE at the root of this distribution.
+#
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+#
+
+set(PAL_TRAIT_RECASTNAVIGATION_SUPPORTED TRUE)
+set(PAL_TRAIT_RECASTNAVIGATION_TEST_SUPPORTED TRUE)
+set(PAL_TRAIT_RECASTNAVIGATION_EDITOR_TEST_SUPPORTED TRUE)
```
Comment (hultonha) [inline Gems/RecastNavigation/Code/Platform/Mac/PAL_mac.cmake:11]: uber nit: EOL as before

File: Gems/RecastNavigation/Code/Source/RecastNavigationEditorSystemComponent.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        RecastNavigationEditorSystemComponent();
+        ~RecastNavigationEditorSystemComponent();
+
+    private:
+        static void GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& provided);
+        static void GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& incompatible);
+        static void GetRequiredServices(AZ::ComponentDescriptor::DependencyArrayType& required);
+        static void GetDependentServices(AZ::ComponentDescriptor::DependencyArrayType& dependent);
+
+        // AZ::Component
```
Comment (hultonha) [inline Gems/RecastNavigation/Code/Source/RecastNavigationEditorSystemComponent.h:36]: uber nit: `AZ::Component overrides ...`

File: Gems/RecastNavigation/Code/Source/RecastNavigationModuleInterface.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        /**
+         * Add required SystemComponents to the SystemEntity.
+         */
+        AZ::ComponentTypeList GetRequiredSystemComponents() const override
+        {
+            return AZ::ComponentTypeList{
+                azrtti_typeid<RecastNavigationSystemComponent>(),
+            };
+        }
```
Comment (hultonha) [inline Gems/RecastNavigation/Code/Source/RecastNavigationModuleInterface.h:41]: nit: comment thing again and also consider moving the implementation to the .cpp (this list can get a lot longer)

File: Gems/RecastNavigation/Code/Platform/Linux/PAL_linux.cmake [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) Contributors to the Open 3D Engine Project.
+# For complete copyright and license terms please see the LICENSE at the root of this distribution.
+#
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+#
+#
+
+set(PAL_TRAIT_RECASTNAVIGATION_SUPPORTED TRUE)
+set(PAL_TRAIT_RECASTNAVIGATION_TEST_SUPPORTED TRUE)
+set(PAL_TRAIT_RECASTNAVIGATION_EDITOR_TEST_SUPPORTED TRUE)
```
[the remark this answers is not in the document]
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Platform/Linux/PAL_linux.cmake]: Fixed.

File: Gems/RecastNavigation/Code/Platform/Mac/PAL_mac.cmake (hunk 4 above)
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Platform/Mac/PAL_mac.cmake]: Fixed.

File: Gems/RecastNavigation/Platform/Windows/windows_gem.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+  "Tags": ["Windows"]
+}
```
[the remark this answers is not in the document]
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Platform/Windows/windows_gem.json]: Fixed.

File: Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+#include <AzCore/Serialization/SerializeContext.h>
+#include <AzCore/Serialization/EditContext.h>
+#include <AzCore/Serialization/EditContextConstants.inl>
+
+namespace RecastNavigation
+{
+    void RecastNavigationSystemComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
```
[the remark this answers is not in the document]
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp]: Fixed.

File: Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            serialize->Class<RecastNavigationSystemComponent, AZ::Component>()
+                ->Version(0)
+                ;
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
```
[the remark this answers is not in the document]
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationSystemComponent.cpp]: Fixed.

File: Gems/RecastNavigation/Code/Source/RecastNavigationModuleInterface.h (hunk 6 above)
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationModuleInterface.h]: Fixed the comment style. This header is included in multiple module files only.

File: Gems/RecastNavigation/Code/Source/RecastNavigationEditorSystemComponent.h (hunk 5 above)
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationEditorSystemComponent.h]: Fixed.

File: Gems/RecastNavigation/Code/Source/RecastNavigationEditorSystemComponent.h [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+#pragma once
+
+#include <RecastNavigationSystemComponent.h>
+
+#include <AzToolsFramework/Entity/EditorEntityContextBus.h>
+
+namespace RecastNavigation
+{
+    /// System component for RecastNavigation editor
```
[the remark this answers is not in the document]
Comment (AMZN-Olex) [reply Gems/RecastNavigation/Code/Source/RecastNavigationEditorSystemComponent.h]: Fixed.

**Review by SergeyAMZN — APPROVED** [review]

**Pull request merged**